### PR TITLE
filesystem: abort filesystem bucket operations if the context has been cancelled

### PR DESCRIPTION
**Pull request by charleskorn**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Changed all of the `filesystem.Bucket` methods to return early if the context has been cancelled.

This makes testing cancellation behaviour of consuming applications easier (see https://github.com/grafana/mimir/pull/4061 for an example), and brings the behaviour of the filesystem bucket into line with the other providers.

## Verification

I've added tests for cancellation behaviour in `filesystem_test.go`, and the existing acceptance tests should catch any regressions in existing behaviour.

### CODE REVIEW

File: providers/filesystem/filesystem.go [hunk 1]
```diff
@@ -53,6 +53,12 @@ func NewBucket(rootDir string) (*Bucket, error) {
 // Iter calls f for each entry in the given directory. The argument to f is the full
 // object name including the prefix of the inspected directory.
 func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error, options ...objstore.IterOption) error {
+	select {
```
Comment (pracucci) [inline providers/filesystem/filesystem.go:56]: [nit] I think you can simplify this into:

```
if ctx.Err() != nil {
  return ctx.Err()
}
```

Comment (charleskorn) [reply providers/filesystem/filesystem.go]: Good point, resolved in 666b242.

**Review by pracucci — APPROVED** [review]
FWIW, LGTM :)

**Review by douglascamata — APPROVED** [review]
LGTM ✅

**Review by yeya24 — COMMENTED** [review]
Thanks!

**Review by yeya24 — APPROVED** [review]

**Review by PhilipGough — APPROVED** [review]

**Pull request merged**